### PR TITLE
Add missing docs regarding diagonal

### PR DESCRIPTION
**Pull request by jorgensd**
Also add some notes about when diagonal is set.

### CODE REVIEW

File: python/dolfinx/fem/assemble.py [hunk 1]
```diff
@@ -245,6 +246,9 @@ def assemble_matrix(
     Note:
         The returned matrix is not finalised, i.e. ghost values are not
         accumulated.
+
+    Note:
+        Diagonal entries are only inserted for matrices with the same trial and test space.
```
Comment (jhale) [inline python/dolfinx/fem/assemble.py:251]: Notes are more asides. As this is a critical point I would add it to the argument as 'for constrained degrees of freedom belonging to the same trial and test space.'

**Review by jhale — CHANGES_REQUESTED** [review]

File: python/dolfinx/fem/petsc.py [hunk 2]
```diff
@@ -431,6 +431,9 @@ def assemble_matrix(
         The returned matrix is not 'assembled', i.e. ghost contributions
         have not been communicated.
 
+    Note:
```
Comment (jhale) [inline python/dolfinx/fem/petsc.py:434]: Same comment.

**Review by jhale — DISMISSED** [review]

File: python/dolfinx/fem/petsc.py [hunk 3]
```diff
@@ -431,11 +431,14 @@ def assemble_matrix(
...
+        Diagonal entries are only inserted for matrices with the same trial and test space.
+
```
Comment (jorgensd) [inline python/dolfinx/fem/petsc.py:436]:
```suggestion
```

**Review by garth-wells — APPROVED** [review]

**Comment by garth-wells** [comment]
@jhale could you approve this PR if you're happy with it?

**Pull request merged**